doc/adr/0007-support-reference-inputs-in-constraint-library.rst

```diff
@@ -0,0 +1,47 @@
+.. _support_reference_inputs_in_constraint_library:
+
+ADR 7: Support reference inputs in constraint library
+=====================================================
+
+Date: 2022-08-09
+
+Authors
+-------
+
+koslambrou <konstantinos.lambrou@iohk.io>
+
+Status
+------
+
+Draft
+
+Context
+-------
+
+After the Vasil HF, the Cardano blockchain will support `reference` inputs by adding a new field in the transaction data type.
+With reference inputs, transactions can take a look at UTXOs without actually spending them.
+
+Thus, we need to adapt our transaction constraint data type (`TxConstraints`) to support referencing UTXOs.
+
+Decision
+--------
+
+* We will add the data constuctor `MustReferenceOutput TxOutRef` to the `TxConstraints` data type.
+
+* The PlutusV1 on-chain implementation of this new constraint will simply return `False`.
+  However, `cardano-ledger` throws a phase-1 validation error if transactions that use the some of the new features (reference inputs, inline datums and reference scripts) try to execute PlutusV1 scripts.
+  See the `Babbage era ledger specification <https://hydra.iohk.io/job/Cardano/cardano-ledger/specs.babbage-ledger/latest/download-by-type/doc-pdf/babbage-changes>`_.
+  Therefore, this on-chain implementation should **never** result in a phase-2 validation error.
+
+* The PlutusV2 on-chain implementation of this new constraint will check that the provided `TxOutRef` is part of the `ScriptContext`'s reference inputs.
+
+Argument
+--------
+
+At first glance, we might think that we need two data constructors for reference inputs such as `MustReferencePubKeyOutput` and `MustReferenceScriptOutput` in contrast to the existing `MustSpendPubKeyOutput` and `MustSpendScriptOutput` constraints.
+However, we do not need to make the distinction between public key outputs and script outputs because we're not spending the output, therefore, we don't need to provide a redeemer nor the actual script as a witness to the transaction input.
```

Comment:
Yeah exactly. That's what I'm saying. We can just have `MustReferenceOutput TxOutRef` because we don't need to know the difference between a pubkey and script output.